spec-docs/function-types.md

```diff
@@ -0,0 +1,304 @@
+# Function Types in Kotlin on JVM
+
+## Goals
+
+* Get rid of 23 hardwired physical function classes. One of the problems with them is that they should be effectively duplicated in reflection which means a lot of physical classes in the runtime.
+* Make extension functions coercible to normal functions (with an extra parameter), so that it's possible to do `listOfStrings.map(String::length)`
+* Allow functions with more than 23 parameters, theoretically any number of parameters (in practice 255 on JVM).
+* At the same time, allow to implement Kotlin functions easily from Java: `new Function2() { ... }` and overriding `invoke` only would be the best.
+Enabling SAM conversions on Java 8 would also be terrific.
+
+## Brief solution overview
+
+* Treat extension functions almost like non-extension functions with one extra parameter, allowing to use them almost interchangeably.
+* Introduce a physical class `Function` and unlimited number of *fictitious* (synthetic) classes `Function0`, `Function1`, ... in the compiler front-end
+* On JVM, introduce `Function0`..`Function22`, which are optimized in a certain way,
+and `FunctionN` for functions with 23+ parameters.
+When passing a lambda to Kotlin from Java, one will need to implement one of these interfaces.
+* Also on JVM (under the hood) add abstract `FunctionImpl` which implements all of `Fun0`..`Fun22` and `FunN`
+(throwing exceptions), and which knows its arity.
+Kotlin lambdas are translated to subclasses of this abstract class, passing the correct arity to the super constructor.
+* Provide a way to get arity of an arbitrary `Function` object (pretty straightforward).
+* Hack `is/as Function5` on any numbered function in codegen (and probably `KClass.cast()` in reflection) to check against `Function` and its arity.
```

Comment:
So, is it `Function` or `FunctionImpl` who knows the function arity? For which type are we going to check in `is` operator?